WebApp/autoreduce_webapp/selenium_tests/tests/test_run_summary_integration.py

```diff
@@ -0,0 +1,187 @@
+        cls.database_client = DatabaseClient()
+        cls.database_client.connect()
+        try:
+            cls.queue_client, cls.listener = main()
+        except ConnectionException as err:
+            raise RuntimeError("Could not connect to ActiveMQ - check you credentials. If running locally check that "
+                               "ActiveMQ is running and started by `python setup.py start`") from err
+
+        cls.instrument_name = "TestInstrument"
+        cls.rb_number = 1234567
+        cls.run_number = 99999
+
+        cls.data_archive = DataArchive([cls.instrument_name], 21, 21)
+        cls.data_archive.create()
+        cls.data_archive.add_reduce_vars_script(cls.instrument_name,
+                                                """standard_vars={"variable1":"test_variable_value_123"}""")
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.queue_client.disconnect()
+        cls.database_client.disconnect()
+        cls.data_archive.delete()
+        super().tearDownClass()
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.page = RunSummaryPage(self.driver, self.instrument_name, 99999, 0)
+        self.page.launch()
+        # clicks the toggle to show the rerun panel, otherwise the buttons in the form are non interactable
+        self.page.toggle_button.click()
+
+    def _find_run_in_database(self):
+        """
+        Find a ReductionRun record in the database
+        This includes a timeout to wait for several seconds to ensure the database has received
+        the record in question
+        :return: The resulting record
+        """
+        instrument = db.get_instrument(self.instrument_name)
+        return instrument.reduction_runs.filter(run_number=self.run_number)
+
+    def wait_for_result(self):
+        """Waits until the queue listener has finished processing the current message"""
+        # forces the is_processing to return True so that the listener has time to actually start processing the message
+        self.listener._processing = True  #pylint:disable=protected-access
+        while self.listener.is_processing_message():
+            time.sleep(0.5)
+
+        # Get Result from database
+        results = self._find_run_in_database()
+
+        assert results
+        return results
+
+    def submit_after_reset(self):
+        """
+        Submit after a reset button has been clicked.
+
+        Sticks the submission in a loop in case the first time doesn't work. The reason
+        it may not work is that resetting actually swaps out the whole form using JS, which
+        replaces ALL the elements and triggers a bunch of DOM re-renders/updates, and that isn't fast.
+        """
+        while str(self.run_number) in self.driver.current_url:
+            # NOTE that we MUST wait first and then try to submit, otherwise the processing
+            # may be finished before wait_for_result is called, causing it to be stuck forever
+            time.sleep(1)
```

Comment:
Can you use a selenium wait until expected condition here instead of a sleep?

Reply:
I'll look into that! These potentially infinite while's already caused the tests to deadlock for 6 hours, so it would be good to have the wait until with a timeout